### ISSUE TITLE
error in refreshing the rules

### ISSUE THREAD

**Issue by mlodic** [rseichter/fangfrisch#17]
Hello!

first of all, thank you for this work!

While trying to integrate this tool with [IntelOwl](https://github.com/intelowlproject/IntelOwl), I found a problem that always happen when I launch `fangfrisch refresh`

```
intelowl_malware_tools_analyzers | INFO: /var/lib/clamav/junk.ndb updated (6989248 bytes)
intelowl_malware_tools_analyzers | INFO: /var/lib/clamav/jurlbl.ndb updated (830565 bytes)
intelowl_malware_tools_analyzers | INFO: /var/lib/clamav/jurlbla.ndb updated (88784 bytes)
intelowl_malware_tools_analyzers | Traceback (most recent call last):
intelowl_malware_tools_analyzers |   File "/usr/local/bin/fangfrisch", line 8, in <module>
intelowl_malware_tools_analyzers |     sys.exit(main())
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/fangfrisch/__main__.py", line 64, in main
intelowl_malware_tools_analyzers |     ClamavRefresh(args).refresh_all()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/fangfrisch/refresh.py", line 142, in refresh_all
intelowl_malware_tools_analyzers |     if self.refresh(ci):
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/fangfrisch/refresh.py", line 134, in refresh
intelowl_malware_tools_analyzers |     RefreshLog.update(ci, digest.data)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/fangfrisch/db.py", line 190, in update
intelowl_malware_tools_analyzers |     entry: RefreshLog = _query_url(ci.url, session)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/fangfrisch/db.py", line 239, in _query_url
intelowl_malware_tools_analyzers |     return session.query(RefreshLog).filter(RefreshLog.url == url).first()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2752, in first
intelowl_malware_tools_analyzers |     return self.limit(1)._iter().first()  # type: ignore
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2855, in _iter
intelowl_malware_tools_analyzers |     result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2229, in execute
intelowl_malware_tools_analyzers |     return self._execute_internal(
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2114, in _execute_internal
intelowl_malware_tools_analyzers |     conn = self._connection_for_bind(bind)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1981, in _connection_for_bind
intelowl_malware_tools_analyzers |     return trans._connection_for_bind(engine, execution_options)
intelowl_malware_tools_analyzers |   File "<string>", line 2, in _connection_for_bind
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/state_changes.py", line 137, in _go
intelowl_malware_tools_analyzers |     ret_value = fn(self, *arg, **kw)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1108, in _connection_for_bind
intelowl_malware_tools_analyzers |     conn = bind.connect()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3245, in connect
intelowl_malware_tools_analyzers |     return self._connection_cls(self)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
intelowl_malware_tools_analyzers |     self._dbapi_connection = engine.raw_connection()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3269, in raw_connection
intelowl_malware_tools_analyzers |     return self.pool.connect()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 455, in connect
intelowl_malware_tools_analyzers |     return _ConnectionFairy._checkout(self)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 1270, in _checkout
intelowl_malware_tools_analyzers |     fairy = _ConnectionRecord.checkout(pool)
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 719, in checkout
intelowl_malware_tools_analyzers |     rec = pool._do_get()
intelowl_malware_tools_analyzers |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 157, in _do_get
intelowl_malware_tools_analyzers |     raise exc.TimeoutError(
intelowl_malware_tools_analyzers | sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)

```

At one point, during download, SQLAlchemy breaks. 
I think that by incrementing the `pool_size` and `max_overflow` values  (see [doc](https://docs.sqlalchemy.org/en/20/errors.html#connections-and-transactions)) it would fix the problem.

I can open a really little PR for this if you like

**Comment by rseichter** [comment]
Thanks for the report. I will need more information to investigate this, though. You did not mention your platform, Python module versions used, and how to reproduce the issue, so I cannot do much at this point. 😉

**Comment by mlodic** [comment]
#18 solves the issue if you like to merge it. As I mentioned, it is just a matter of SQLAlchemy configuration

**Comment by rseichter** [comment]
Did you read my previous comment? You did not provide enough information. I am not going to make any changes to Fangfrisch without even being able to reproduce an issue that has only ever been reported by you and nobody else. That should not come as a surprise. It could be an SQL driver issue, an SQLalchemy problem, a problem of your computer, or whatever.

**Comment by mlodic** [comment]
I am just trying to help the community that is using this tool but it does not seem welcome. There no need to be rude. Now I get why you are the single developer in this project. 🤷🏻‍♂️ 

Anyway, this is my setup:

I am installing fangfrisch inside an image run with this Dockerfile: https://github.com/intelowlproject/IntelOwl/blob/master/integrations/malware_tools_analyzers/Dockerfile
there is python 3.8-slim image, Debian bullseye.

Then I have already told you how to reproduce the issue, which is just by running `fangfrisch refresh`. The configuration I used is this: https://github.com/intelowlproject/IntelOwl/blob/develop/integrations/malware_tools_analyzers/clamav/fangfrisch.conf

**Comment by rseichter** [comment]
Nothing about me asking for additional information was rude, you simply chose not to answer me and I had to ask once again. Before accusing others of being rude, you may want to revisit how one submits bug reports.

Also, as I mentioned, nobody but you has reported this issue so far, and I have not been able to reproduce it myself. Hence, I told you that I won't make changes. Fangfrisch has been working fine for me and others since 2020, so of course I am cautious and don't add code modifications willy-nilly to deal with a symptom I have not yet seen occur, without first trying to find the root cause for a possible, as of now unconfirmed, problem first. That's just common sense.

**Comment by gchamon** [comment]
Happening to me too, using docker (image python:latest) and the configuration provided by @mlodic

I just ran a container using the base image and followed the installation instructions on the site.

![image](https://user-images.githubusercontent.com/9471861/219398067-08c4c5e4-0553-41bd-8e20-140a37cd6b5b.png)

@rseichter maybe we could reopen the issue?

**Comment by gchamon** [comment]
I can confirm that applying patch https://github.com/rseichter/fangfrisch/pull/18 fixes the issue

**Comment by gchamon** [comment]
no luck trying to work around the issue.

@rseichter a safer approach to this problem would be to expose pool size and max overflow as configurations for `fangfrisch.conf`, for instance under `DEFAULT.sqlalchemy_pool_size` and `DEFAULT.sqlalchemy_max_overflow`. In case those configurations are omitted, the engine is created using default settings. This way there is no possibility of regression for any user.

**Comment by rseichter** [comment]
@gchamon Am I correct to assume that by „installation instructions on the site“ you mean the Fangfrisch online documentation? Also, could you please provide a Python module list, including the exact version of SQLalchemy, and attach your fangfrisch.conf here?

**Comment by gchamon** [comment]
> Am I correct to assume that by „installation instructions on the site“ you mean the Fangfrisch online documentation?

Yes! By using a virtualenv, giving the correct permissions to the folders mentioned in the site etc... basically this: https://rseichter.github.io/fangfrisch/#_installation

> Python module list, including the exact version of SQLalchemy

```
certifi==2022.12.7
charset-normalizer==3.0.1
fangfrisch==1.5.0
greenlet==2.0.2
idna==3.4
requests==2.28.2
SQLAlchemy==2.0.3
typing_extensions==4.5.0
urllib3==1.26.14
```

> fangfrisch.conf

https://github.com/gchamon/fangfrisch-queuepool-poc/blob/28c17e7cec5dd267e11381bc449dbc85d05023a2/fangfrisch.conf

-----

I have also created a working proof-of-concept of the issue using docker (for reproducibility and isolation) which can be found here: https://github.com/gchamon/fangfrisch-queuepool-poc

Instructions in the README

**Comment by rseichter** [comment]
@gchamon Now *that* is information I can work with. Thank you for taking the time to collect the information and even provide an isolated Git project with which I can finally reproduce the problem. I forked the POC and will look into this issue.

**Comment by gchamon** [comment]
@rseichter 1.6.0 solves the issue!

**Comment by rseichter** [comment]
Increasing the pool size would only have been like a mere coat of paint over an underlying flaw in handling database sessions. It took your information and additional testing on my end to realise that my previous code was too "optimistic" when it came to long-running DB connections. I hope the new way of handling sessions will be robust.

**Comment by rseichter** [comment]
Closing this issue as fixed with Fangfrisch release 1.6.0. Thanks again, @gchamon .

**Comment by gchamon** [comment]
@rseichter awesome! Nice job!
